test/Microsoft.Extensions.Logging.EventSource.Test/EventSourceLoggerTest.cs

```diff
@@ -0,0 +1,581 @@
+                    "E8FM", "E8MSG", "E8JS");
+#endif
+            }
+        }
+
+        [Fact]
+        public void Logs_Nothing_IfNotEnabled()
+        {
+            TestEventListener.Settings.Keywords = EventKeywords.None;
+            TestEventListener.Settings.FilterSpec = null;
+            TestEventListener.Settings.Level = default(EventLevel);
+            using (var testListener = new TestEventListener())
+            {
+                var factory = new LoggerFactory();
+                // No call to factory.AddEventSourceLogger();
+
+                LogStuff(factory);
+
+                VerifyEvents(testListener); // No verifiers = 0 events expected
+            }
+        }
+
+        [Fact]
+        public void Logs_OnlyFormattedMessage_IfKeywordSet()
+        {
+            TestEventListener.Settings.Keywords = LoggingEventSource.Keywords.FormattedMessage;
+            TestEventListener.Settings.FilterSpec = null;
+            TestEventListener.Settings.Level = EventLevel.Verbose;
+            using (var testListener = new TestEventListener())
+            {
+                var factory = new LoggerFactory();
+                factory.AddEventSourceLogger();
+
+                LogStuff(factory);
+
+                VerifyEvents(testListener,
+                    "E1FM",
+                    // Second event is omitted because default LogLevel == Debug 
+                    "E3FM", 
+                    "OuterScopeStart",
+                    "E4FM", 
+                    "E5FM", 
+                    "InnerScopeStart",
+                    "E6FM", 
+                    "InnerScopeStop",
+                    "E7FM", 
+                    "OuterScopeStop",
+                    "E8FM");
+            }
+        }
+
+        [Fact]
+        public void Logs_OnlyJson_IfKeywordSet()
+        {
+            TestEventListener.Settings.Keywords = LoggingEventSource.Keywords.JsonMessage;
+            TestEventListener.Settings.FilterSpec = null;
+            TestEventListener.Settings.Level = EventLevel.Verbose;
+            using (var testListener = new TestEventListener())
+            {
+                var factory = new LoggerFactory();
+                factory.AddEventSourceLogger();
+
+                LogStuff(factory);
+
+                VerifyEvents(testListener,
+                    "E1JS",
+                    // Second event is omitted because default LogLevel == Debug 
+                    "E3JS",
+                    "OuterScopeJsonStart",
+                    "E4JS",
+                    "E5JS",
+                    "InnerScopeJsonStart",
+                    "E6JS",
+                    "InnerScopeJsonStop",
+                    "E7JS",
+                    "OuterScopeJsonStop",
+                    "E8JS");
+            }
+        }
+
+        [Fact]
+        public void Logs_OnlyMessage_IfKeywordSet()
+        {
+            TestEventListener.Settings.Keywords = LoggingEventSource.Keywords.Message;
```

Comment:
This is explained in the comment inside TestEventListener. It is a workaround. Basically, there is a bug in EventListener implementation, which causes it to call into one of the overloads (OnEventSourceCreated) before the instance is fully constructed. So we cannot use TestEventListener constructor parameters to initialize the instance; the initialization parameters must be supplied externally.